File: ios/HangulEngine/Sources/HangulEngine/HangulTables.swift
enum HangulTables {
  static let leading: [Character] = Array("ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ")
  static let medial: [Character] = Array("ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ")
  static let trailing: [Character?] = [
    nil, "ㄱ", "ㄲ", "ㄳ", "ㄴ", "ㄵ", "ㄶ", "ㄷ", "ㄹ", "ㄺ", "ㄻ", "ㄼ", "ㄽ", "ㄾ",
    "ㄿ", "ㅀ", "ㅁ", "ㅂ", "ㅄ", "ㅅ", "ㅆ", "ㅇ", "ㅈ", "ㅊ", "ㅋ", "ㅌ", "ㅍ", "ㅎ",
  ]

  static let compoundMedial: [JamoPair: Character] = [
    JamoPair("ㅗ", "ㅏ"): "ㅘ",
    JamoPair("ㅗ", "ㅐ"): "ㅙ",
    JamoPair("ㅗ", "ㅣ"): "ㅚ",
    JamoPair("ㅜ", "ㅓ"): "ㅝ",
    JamoPair("ㅜ", "ㅔ"): "ㅞ",
    JamoPair("ㅜ", "ㅣ"): "ㅟ",
    JamoPair("ㅡ", "ㅣ"): "ㅢ",
  ]

  static let compoundTrailing: [JamoPair: Character] = [
    JamoPair("ㄱ", "ㅅ"): "ㄳ",
    JamoPair("ㄴ", "ㅈ"): "ㄵ",
    JamoPair("ㄴ", "ㅎ"): "ㄶ",
    JamoPair("ㄹ", "ㄱ"): "ㄺ",
    JamoPair("ㄹ", "ㅁ"): "ㄻ",
    JamoPair("ㄹ", "ㅂ"): "ㄼ",
    JamoPair("ㄹ", "ㅅ"): "ㄽ",
    JamoPair("ㄹ", "ㅌ"): "ㄾ",
    JamoPair("ㄹ", "ㅍ"): "ㄿ",
    JamoPair("ㄹ", "ㅎ"): "ㅀ",
    JamoPair("ㅂ", "ㅅ"): "ㅄ",
  ]

  static let splitMedial: [Character: JamoPair] = Dictionary(
    uniqueKeysWithValues: compoundMedial.map { ($0.value, $0.key) }
  )
  static let splitTrailing: [Character: JamoPair] = Dictionary(
    uniqueKeysWithValues: compoundTrailing.map { ($0.value, $0.key) }
  )

  static let shiftedJamo = Set(Array("ㄲㄸㅃㅆㅉㅒㅖ"))
  static let allowedLiteralPunctuation = Set(Array(" .,!?…'\"()-·~♡♥。！？"))

  static let leadingIndex = Dictionary(
    uniqueKeysWithValues: leading.enumerated().map { ($0.element, $0.offset) })
  static let medialIndex = Dictionary(
    uniqueKeysWithValues: medial.enumerated().map { ($0.element, $0.offset) })
  static let trailingIndex = Dictionary(
    uniqueKeysWithValues: trailing.enumerated().compactMap { index, character in
      character.map { ($0, index) }
    }
  )
}

struct JamoPair: Hashable, Sendable {
  let first: Character
  let second: Character

  init(_ first: Character, _ second: Character) {
    self.first = first
    self.second = second
  }
}

File: android/core/hangul/src/main/kotlin/app/piyokey/core/hangul/HangulTables.kt
package app.piyokey.core.hangul

internal data class JamoPair(
  val first: Char,
  val second: Char,
)

internal object HangulTables {
  val leading = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅆㅇㅈㅉㅊㅋㅌㅍㅎ".toList()
  val medial = "ㅏㅐㅑㅒㅓㅔㅕㅖㅗㅘㅙㅚㅛㅜㅝㅞㅟㅠㅡㅢㅣ".toList()
  val trailing: List<Char?> = listOf(
    null,
    'ㄱ', 'ㄲ', 'ㄳ', 'ㄴ', 'ㄵ', 'ㄶ', 'ㄷ', 'ㄹ', 'ㄺ', 'ㄻ', 'ㄼ', 'ㄽ', 'ㄾ',
    'ㄿ', 'ㅀ', 'ㅁ', 'ㅂ', 'ㅄ', 'ㅅ', 'ㅆ', 'ㅇ', 'ㅈ', 'ㅊ', 'ㅋ', 'ㅌ', 'ㅍ', 'ㅎ',
  )

  val compoundMedial = mapOf(
    JamoPair('ㅗ', 'ㅏ') to 'ㅘ',
    JamoPair('ㅗ', 'ㅐ') to 'ㅙ',
    JamoPair('ㅗ', 'ㅣ') to 'ㅚ',
    JamoPair('ㅜ', 'ㅓ') to 'ㅝ',
    JamoPair('ㅜ', 'ㅔ') to 'ㅞ',
    JamoPair('ㅜ', 'ㅣ') to 'ㅟ',
    JamoPair('ㅡ', 'ㅣ') to 'ㅢ',
  )

  val compoundTrailing = mapOf(
    JamoPair('ㄱ', 'ㅅ') to 'ㄳ',
    JamoPair('ㄴ', 'ㅈ') to 'ㄵ',
    JamoPair('ㄴ', 'ㅎ') to 'ㄶ',
    JamoPair('ㄹ', 'ㄱ') to 'ㄺ',
    JamoPair('ㄹ', 'ㅁ') to 'ㄻ',
    JamoPair('ㄹ', 'ㅂ') to 'ㄼ',
    JamoPair('ㄹ', 'ㅅ') to 'ㄽ',
    JamoPair('ㄹ', 'ㅌ') to 'ㄾ',
    JamoPair('ㄹ', 'ㅍ') to 'ㄿ',
    JamoPair('ㄹ', 'ㅎ') to 'ㅀ',
    JamoPair('ㅂ', 'ㅅ') to 'ㅄ',
  )

  val splitMedial = compoundMedial.entries.associate { (pair, compound) -> compound to pair }
  val splitTrailing = compoundTrailing.entries.associate { (pair, compound) -> compound to pair }

  val shiftedJamo = "ㄲㄸㅃㅆㅉㅒㅖ".toSet()
  val allowedLiteralPunctuation = " .,!?…'\"()-·~♡♥。！？".toSet()

  val leadingIndex = leading.withIndex().associate { (index, character) -> character to index }
  val medialIndex = medial.withIndex().associate { (index, character) -> character to index }
  val trailingIndex = trailing.withIndex().mapNotNull { (index, character) ->
    character?.let { it to index }
  }.toMap()
}
